### DISCUSSION TITLE
Different behaviour of keys() function depending on how you select data

### DISCUSSION THREAD

**Discussion by aranvir** [sqlalchemy/sqlalchemy#9470] [Usage Questions] [answered]
Hi, I struggled a bit with the [engine.Result.keys()](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.keys) method because it is behaving differently depending on how I select my data.

I worked with Python 3.10.7 and sqlalchemy 2.0.3.

The model details are not important here, so let's just assume a very simple ORM model like this:
```python
class Base(DeclarativeBase):
    pass


class MockData(Base):
    __tablename__ = "mock"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(30))
```

There are two ways for me to query the "total" data. With the explicit query, I specify each column of the table. With the implicit query, I just trust the library to figure it out. The implicit query is what I would have assumed to be `SELECT * FROM mock`.

```python
def get_all_explicit():
    with Session(ENGINE) as session:
        stmt = select(MockData.id, MockData.name)
        result = session.execute(stmt)
        keys = result.keys()
        values = result.all()

def get_all_implicit():
    with Session(ENGINE) as session:
        stmt = select(MockData)
        result = session.execute(stmt)
        keys = result.keys()
        values = result.all()
```

Now regarding the `keys()` function. For the explicit call, I get what I expected: An iterator over the column names of my query. For the implicit call, I was surprised: It is an iterator over all table names queried.

I'm sure there is a reason for this and it is even documented so I should not be surprised in the first place... However, I find this highly unintuitive and it threw me completely off guard. Since I ran this example with `create_engine(..., echo=True)` I saw that for both select statements the query was the exact same. So, the engine queries the columns explicitly in both cases with `SELECT mock.id, mock.name FROM mock`. But then somewhere in the post-processing there has to be a decision on what `keys()` returns...

My question now is: How can I make an implicit query so that `keys()` gives me the column names instead of the table name? I hope there is some parameter for`options()` or `exectuion_options()` or another method to format this output. I couldn't find it though while consulting the documentation..

**Comment by CaselIT** [comment, marked as answer]
Hi,

Keys is consistent with what you get inside the results.
Since you are using the orm, in the first call you get rows of two elements, id and name; in the second call you get back a row of a single element that are instance of MockData. Keys reflect that.

If you don't what to use the orm, either select from `MockData.__table__` or use a connection instead of a session

**Reply by aranvir** [reply]
Thanks a lot!
